### PR TITLE
Improve shutdown to clear BLE API and not just SD

### PR DESCRIPTION
**Pull request by andresag01**
Improve the shutdown functionality, such that a call to ble.shutdown() from
the user application clears the API and nRF5x state and NOT only the
SoftDevice. To achieve this the following changes are introduced:

* Add a protected member cleanup() to nRF5xGap, nRF5xGattClient,
nRF5xGattServer, nRF5xSecurityManager and nRF5xServiceDiscovery.
* Modify the shutdown() implementation in nRF5xn such that it also calls the
static member shutdown() exposed by the BLE API in Gap.h, SecurityManager.h,
GattClient.h and GattServer.h.
* Modify nRF5xGattClient, nRF5xGattServer and nRF5xSecurityManager
classes so that they dynamically create their respective objects only if
needed. Previously the GattClient, GattServer and SecurityManager objects were
declared as static, which means that they were always present even though they
were not always needed. This increases memory consumption unnecessarily.
Furthermore, pointers to the object instances are stored in static members of
the classes as specified by the BLE API base classes. This ensures that
calls to shutdown do not require calls to getInstance() functions that would
otherwise result in undesired memory allocations.
* nRF5xGap object is always needed, so this remains allocated statically. But
the reference in Gap is pointed to this object.

The shutdown procedure is as follows:

1. The user calls ble.shutdown() which executes the code in nRF5xn::shutdown()
1. The SoftDevice is shutdown
1. The static members of Gap.h, SecurityManager.h, GattClient.h and
GattServer.h are called to clean up their own state.

If at any point an error occur during the last step, BLE_ERROR_INVALID_STATE is
returned.

**NOTE:** This enhancement also requires changes to the [ble module](https://github.com/ARMmbed/ble)
@rgrover @pan- @LiyouZhou

### CODE REVIEW

**Comment by andresag01** [comment]
@rgrover @pan- I have made changes according to comments in [this pull request](https://github.com/ARMmbed/ble/pull/141).

File: source/nRF5xn.cpp [hunk 1]
```diff
@@ -104,16 +121,60 @@ ble_error_t nRF5xn::init(BLE::InstanceID_t instanceID, FunctionPointerWithContex
...
+     */
     if(softdevice_handler_sd_disable() != NRF_SUCCESS) {
         return BLE_STACK_BUSY;
     }
 
+
+    /* Shutdown the BLE API and nRF51 glue code */
+    if (gattServerInstance != NULL &&
+        gattServerInstance->reset() != BLE_ERROR_NONE) {
+        return BLE_ERROR_INVALID_STATE;
```
Comment (pan-) [inline source/nRF5xn.cpp]: Could you return the error code instead of BLE_ERROR_INVALID_STATE ?

File: source/nRF5xn.cpp [hunk 2]
```diff
@@ -104,16 +121,60 @@ ble_error_t nRF5xn::init(BLE::InstanceID_t instanceID, FunctionPointerWithContex
...
+    }
+
+    if (securityManagerInstance != NULL &&
+        securityManagerInstance->reset() != BLE_ERROR_NONE) {
+        return BLE_ERROR_INVALID_STATE;
```
Comment (pan-) [inline source/nRF5xn.cpp]: Could you return the error code instead of BLE_ERROR_INVALID_STATE ?

File: source/nRF5xn.cpp [hunk 3]
```diff
@@ -104,16 +121,60 @@ ble_error_t nRF5xn::init(BLE::InstanceID_t instanceID, FunctionPointerWithContex
...
+    }
+
+    /* S110 does not support BLE client features, nothing to reset. */
+#if !defined(TARGET_MCU_NRF51_16K_S110) && !defined(TARGET_MCU_NRF51_32K_S110)
+    if (gattClientInstance != NULL &&
+        gattClientInstance->reset() != BLE_ERROR_NONE) {
+        return BLE_ERROR_INVALID_STATE;
```
Comment (pan-) [inline source/nRF5xn.cpp]: Could you return the error code instead of BLE_ERROR_INVALID_STATE ?

File: source/nRF5xn.cpp [hunk 4]
```diff
@@ -104,16 +121,60 @@ ble_error_t nRF5xn::init(BLE::InstanceID_t instanceID, FunctionPointerWithContex
...
+    }
+#endif
+
+    if (gapInstance != NULL &&
+        gapInstance->reset() != BLE_ERROR_NONE) {
+        return BLE_ERROR_INVALID_STATE;
```
Comment (pan-) [inline source/nRF5xn.cpp]: Could you return the error code instead of BLE_ERROR_INVALID_STATE ?

File: source/nRF5xn.h [hunk 5]
```diff
@@ -41,32 +41,153 @@ class nRF5xn : public BLEInstanceBase
...
 
+public:
+    static nRF5xn& Instance(BLE::InstanceID_t instanceId);
+
 private:
     bool              initialized;
     BLE::InstanceID_t instanceID;
+
+private:
+    static nRF5xGap _gapInstance; /**< Gap instance whose reference is returned from a call to
```
Comment (rgrover) [inline source/nRF5xn.h]: this shouldn't be `static`, or am I missing something?

File: source/nRF5xn.h [hunk 6]
```diff
@@ -41,32 +41,153 @@ class nRF5xn : public BLEInstanceBase
...
+                                   * getGap(). Unlike the GattClient, GattServer and
+                                   * SecurityManager, Gap is always needed in a BLE application.
+                                   * Therefore it is allocated statically. */
+
+private:
+    mutable nRF5xGap             *gapInstance;             /**< Pointer to the Gap object instance.
```
Comment (rgrover) [inline source/nRF5xn.h]: do we need to hang on this pointer?

File: source/nRF5xn.h (hunk 5 above)
Comment (andresag01) [inline source/nRF5xn.h]: This does not need to be static, but it used to be like this before the shutdown changes were introduced, so I left it the same way. I can also make it a member of the class, there is no problem with this.

File: source/nRF5xn.h (hunk 6 above)
Comment (andresag01) [inline source/nRF5xn.h]: Indeed this is a bit unnecessary, but I decided to include it for consistency -- there are pointers to the other components: GattClient, GattServer, SecurityManager and Gap. I am happy to remove it as well and modify the getGap() functions to return `&_gapInstance` always.

**Comment by andresag01** [comment]
@rgrover @pan-: Made changes according to comments.

**Comment by andresag01** [comment]
@rgrover made changes to make `_gapInstance` non-static

**Comment by rgrover** [comment]
@andresag01 I'm happy with this. Will wait for @pan-

**Comment by pan-** [comment]
It's alright, I'm looking for the next PR including onShutdown

**Comment by rgrover** [comment]
ble-nrf51822-2.2.7

**Pull request merged**